sdk/python/feast/offline_store.py

```diff
@@ -0,0 +1,100 @@
+from abc import ABC, abstractmethod
+from datetime import datetime
+from typing import Dict, List, Optional
+
+import pyarrow
+
+from feast.entity import Entity
+from feast.feature import Feature
+from feast.feature_view import FeatureView
+
+
+class OfflineStore(ABC):
+    """
+    OfflineStore is a non-user-facing object used for all interaction between Feast and the service used for offline storage of features. Currently BigQuery is supported.
+    """
+
+    @abstractmethod
+    def pull_table(
+        self, feature_view: FeatureView, start_date: datetime, end_date: datetime,
+    ):
+        pass
+
+
+class BigQueryOfflineStore:
+    """
+    BigQueryOfflineStore is a non-user-facing object used for all interaction between Feast and BigQuery.
+    """
+
+    def pull_table(
+        self, feature_view: FeatureView, start_date: datetime, end_date: datetime,
+    ) -> Optional[pyarrow.Table]:
+        from google.cloud.bigquery_storage import BigQueryReadClient, types
+
+        project, dataset, table = feature_view.inputs.table_ref.split(".")
+        client = BigQueryReadClient()
+
+        requested_session = types.ReadSession()
+        requested_session.table = (
+            f"projects/{project}/datasets/{dataset}/tables/{table}"
+        )
+        requested_session.data_format = types.DataFormat.ARROW
+
+        # if we have mapped fields, use the original field names in the call to BigQuery
+        event_timestamp_column = feature_view.inputs.event_timestamp_column
+        fields = (
+            [entity.name for entity in feature_view.entities]
+            + [feature.name for feature in feature_view.features]
+            + [feature_view.inputs.event_timestamp_column]
+        )
+        if feature_view.inputs.field_mapping is not None:
+            reverse_field_mapping = {
+                v: k for k, v in feature_view.inputs.field_mapping.items()
+            }
+            event_timestamp_column = (
+                reverse_field_mapping[event_timestamp_column]
+                if event_timestamp_column in reverse_field_mapping.keys()
+                else event_timestamp_column
+            )
+            fields = [
+                reverse_field_mapping[col]
+                if col in reverse_field_mapping.keys()
+                else col
+                for col in fields
+            ]
+        requested_session.read_options.selected_fields = fields
+        requested_session.read_options.row_restriction = f"{event_timestamp_column} BETWEEN TIMESTAMP('{start_date}') AND TIMESTAMP('{end_date}')"
+
+        parent = f"projects/{project}"
+        session = client.create_read_session(
+            parent=parent, read_session=requested_session, max_stream_count=1,
+        )
+
+        if len(session.streams) == 0:
+            # will sometimes happen, indicates no data
+            return None
+        reader = client.read_rows(session.streams[0].name)
+        table = reader.to_arrow(session)
```

Comment:
Does this mean that we are loading all source data into memory (not paging)?

Reply:
I need to think through paging a little bit more but think it shouldn't be in this function since we'd want to alternate reading a page into memory and writing it to the online store, so I'm deferring implementing paging until we have the full ingestion workflow